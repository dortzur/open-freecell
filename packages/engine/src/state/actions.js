export const MOVE_TO_HOME_CELL = "MOVE_TO_HOME_CELL";
export const MOVE_TO_COLUMN_CELL = "MOVE_TO_COLUMN_CELL";
export const MOVE_TO_FREE_CELL = "MOVE_TO_FREE_CELL";
import {getSubStack, getAvailableMovesSimple, isCellEmpty} from "./selectors";
import {isMovableStack} from "../services/card-stack-service";
import {CELL_TYPES} from "../constants";
export const moveToHomeCell = (cardNotation, cellIndex) => (dispatch, getState) => {
};
export const moveToColumnCell = (cardNotation, cellIndex) => (dispatch, getState) => {
  const state = getState();
  const subStack = getSubStack(state, {cardNotation});
  const availableMoves = getAvailableMovesSimple(state);
  const isMovable = isMovableStack(subStack, availableMoves);

  if (!isMovable) {
    return dispatch({type: MOVE_TO_COLUMN_CELL, payload: new Error("Can't move these cards")});
  }

  return dispatch({type: MOVE_TO_COLUMN_CELL, payload: {subStack, cellIndex}})

};

export const moveToFreeCell = (cardNotation, cellIndex) => (dispatch, getState) => {
  const state = getState();

  const isEmpty = isCellEmpty(state, {cellType: CELL_TYPES.FREE_CELL, cellIndex});
  if (!isEmpty) {
    return dispatch({type: MOVE_TO_FREE_CELL, payload: new Error("Free cell not empty")});
  }
  const subStack = getSubStack(state, {cardNotation});
  if (subStack.length > 1) {
    return dispatch({type: MOVE_TO_FREE_CELL, payload: new Error("Can only move one card to free cell")});
  }

  return dispatch({type: MOVE_TO_FREE_CELL, payload: {subStack, cellIndex}});

};

