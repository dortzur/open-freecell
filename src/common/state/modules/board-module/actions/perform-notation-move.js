import * as types from '../types';
export const performNotationMove = (notation) => ({
  type: types.PERFORM_NOTATION_MOVE,
  payload: { notation },
});
