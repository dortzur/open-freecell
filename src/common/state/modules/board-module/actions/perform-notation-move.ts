import * as types from '../types';

export const performNotationMove = (notation: string) => ({
  type: types.PERFORM_NOTATION_MOVE,
  payload: { notation },
});
