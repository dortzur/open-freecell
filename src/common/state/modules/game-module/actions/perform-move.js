import * as types from '../types';
export const performMove = (notation) => ({
  type: types.PERFORM_MOVE,
  payload: { notation },
});
