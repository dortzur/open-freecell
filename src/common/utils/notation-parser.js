const invariant = require('invariant');
const CELL_NOTATION = ['a', 'b', 'c', 'd'];
const TABLEAU_NOTATION = ['1', '2', '3', '4', '5', '6', '7', '8'];
const FOUNDATION_NOTATION = ['h'];
const getNotationStruct = (notation, index, value) => ({
  notation,
  index,
  value,
});

const parseNotationLetter = (state, notationLetter) => {
  if (CELL_NOTATION.includes(notationLetter)) {
    const index = CELL_NOTATION.indexOf(notationLetter);
    return getNotationStruct(notationLetter, index, state.cells[index]);
  }
  if (TABLEAU_NOTATION.includes(notationLetter)) {
    const index = TABLEAU_NOTATION.indexOf(notationLetter);
    return getNotationStruct(notationLetter, index, state.tableau[index]);
  } else if (FOUNDATION_NOTATION.includes(notationLetter)) {
    return getNotationStruct(notationLetter, null, state.foundation);
  }
  throw new Error('invalid notation');
};
export const parseNotation = (state, notation) => {
  const source = parseNotationLetter(state, notation[0]);
  const target = parseNotationLetter(state, notation[1]);
  invariant(source.value !== state.foundation, 'invalid notation');

  return { source, target };
};
