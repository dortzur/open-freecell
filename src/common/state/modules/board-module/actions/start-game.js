import * as types from '../types';
import { createGameDeck } from '../../../../utils';

export const startGame = (gameNumber) => {
  const gameDeck = createGameDeck(gameNumber);
  return {
    type: types.START_GAME,
    payload: { gameDeck, gameNumber },
  };
};
