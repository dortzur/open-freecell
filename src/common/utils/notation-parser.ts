import { Cell, CELL_TYPES } from './consts';
import { BoardState } from '../state/modules/board-module';
import { getBottomCard, getSuitFoundation } from '../engine/utils';

const invariant = require('invariant');
export const CELL_NOTATION = ['a', 'b', 'c', 'd'];
export const TABLEAU_NOTATION = ['1', '2', '3', '4', '5', '6', '7', '8'];
export const FOUNDATION_NOTATION = ['h'];

export interface NotationStruct {
  notation: string,
  index: number,
  type: CELL_TYPES,
  value: Cell
}

const getNotationStruct = (notation: string, index: number, type: CELL_TYPES, value: Cell): NotationStruct => ({
  notation,
  index,
  value,
  type,
});


export interface Move {
  source: NotationStruct,
  target: NotationStruct
}

const parseNotationLetter = (state: BoardState, notationLetter: string, suit?: string): NotationStruct => {
  if (CELL_NOTATION.includes(notationLetter)) {
    const index = CELL_NOTATION.indexOf(notationLetter);
    return getNotationStruct(notationLetter, index, CELL_TYPES.CELL, [
      ...state.cell[index],
    ]);
  }
  if (TABLEAU_NOTATION.includes(notationLetter)) {
    const index = TABLEAU_NOTATION.indexOf(notationLetter);
    return getNotationStruct(notationLetter, index, CELL_TYPES.TABLEAU, [
      ...state.tableau[index],
    ]);
  } else if (FOUNDATION_NOTATION.includes(notationLetter) && suit) {
    const foundationCell = getSuitFoundation(state, suit);
    const index = state.foundation.indexOf(foundationCell);
    return getNotationStruct(notationLetter, index, CELL_TYPES.FOUNDATION, foundationCell);
  }
  throw new Error('invalid notation');
};

export const parseNotation = (state: BoardState, notation: string): Move => {
  const source = parseNotationLetter(state, notation[0]);

  const bottomCard = getBottomCard(source.value);
  const target = parseNotationLetter(state, notation[1], bottomCard && bottomCard.suit);
  invariant(source.type !== CELL_TYPES.FOUNDATION, 'invalid notation');

  return { source, target };
};



