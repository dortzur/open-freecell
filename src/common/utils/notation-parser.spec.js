import { parseNotation } from './notation-parser';
import { startGame } from '../state/modules/board-module/actions/start-game';
import gameReducer from '../state/modules/board-module/index';
describe('notation-parser', () => {
  let state = null;
  beforeAll(() => {
    state = gameReducer(undefined, startGame(100));
  });

  it('should parse moves to home', function() {
    const result = parseNotation(state, '1h');
    expect(result.source.index).toEqual(0);
    expect(result.source).toMatchSnapshot();

    expect(result.target.index).toEqual(null);
    expect(result.target.value).toBeArray();
    expect(result.target).toMatchSnapshot();
    expect(() => parseNotation(state, 'h1')).toThrow();
  });

  it('should parse moves between tableau cells', function() {
    const result = parseNotation(state, '18');
    expect(result.source.index).toEqual(0);
    expect(result.source).toMatchSnapshot();
    expect(result.target.index).toEqual(7);
    expect(result.target).toMatchSnapshot();
  });

  it('should parse moves to open cell', function() {
    const resultA = parseNotation(state, '1a');
    expect(resultA.target.index).toEqual(0);
    expect(resultA.target.index).toEqual(0);
    expect(resultA.target).toMatchSnapshot();

    const resultB = parseNotation(state, 'b1');
    expect(resultB.source.index).toEqual(1);
    expect(resultB.target.index).toEqual(0);
    expect(resultB.target).toMatchSnapshot();
  });
});
