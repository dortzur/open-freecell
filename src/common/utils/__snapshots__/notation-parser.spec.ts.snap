// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`notation-parser should parse moves between tableau cells 1`] = `
Object {
  "index": 0,
  "notation": "1",
  "type": "tableau",
  "value": Array [
    Object {
      "color": "red",
      "id": "AD",
      "rank": "A",
      "suit": "D",
      "value": 1,
    },
    Object {
      "color": "red",
      "id": "QH",
      "rank": "Q",
      "suit": "H",
      "value": 12,
    },
    Object {
      "color": "black",
      "id": "QS",
      "rank": "Q",
      "suit": "S",
      "value": 12,
    },
    Object {
      "color": "black",
      "id": "6S",
      "rank": "6",
      "suit": "S",
      "value": 6,
    },
    Object {
      "color": "black",
      "id": "3C",
      "rank": "3",
      "suit": "C",
      "value": 3,
    },
    Object {
      "color": "black",
      "id": "2C",
      "rank": "2",
      "suit": "C",
      "value": 2,
    },
    Object {
      "color": "black",
      "id": "8S",
      "rank": "8",
      "suit": "S",
      "value": 8,
    },
  ],
}
`;

exports[`notation-parser should parse moves between tableau cells 2`] = `
Object {
  "index": 7,
  "notation": "8",
  "type": "tableau",
  "value": Array [
    Object {
      "color": "red",
      "id": "TD",
      "rank": "T",
      "suit": "D",
      "value": 10,
    },
    Object {
      "color": "red",
      "id": "AH",
      "rank": "A",
      "suit": "H",
      "value": 1,
    },
    Object {
      "color": "black",
      "id": "AS",
      "rank": "A",
      "suit": "S",
      "value": 1,
    },
    Object {
      "color": "red",
      "id": "7H",
      "rank": "7",
      "suit": "H",
      "value": 7,
    },
    Object {
      "color": "red",
      "id": "3H",
      "rank": "3",
      "suit": "H",
      "value": 3,
    },
    Object {
      "color": "red",
      "id": "2D",
      "rank": "2",
      "suit": "D",
      "value": 2,
    },
  ],
}
`;

exports[`notation-parser should parse moves to home 1`] = `
Object {
  "index": 0,
  "notation": "1",
  "type": "tableau",
  "value": Array [
    Object {
      "color": "red",
      "id": "AD",
      "rank": "A",
      "suit": "D",
      "value": 1,
    },
    Object {
      "color": "red",
      "id": "QH",
      "rank": "Q",
      "suit": "H",
      "value": 12,
    },
    Object {
      "color": "black",
      "id": "QS",
      "rank": "Q",
      "suit": "S",
      "value": 12,
    },
    Object {
      "color": "black",
      "id": "6S",
      "rank": "6",
      "suit": "S",
      "value": 6,
    },
    Object {
      "color": "black",
      "id": "3C",
      "rank": "3",
      "suit": "C",
      "value": 3,
    },
    Object {
      "color": "black",
      "id": "2C",
      "rank": "2",
      "suit": "C",
      "value": 2,
    },
    Object {
      "color": "black",
      "id": "8S",
      "rank": "8",
      "suit": "S",
      "value": 8,
    },
  ],
}
`;

exports[`notation-parser should parse moves to home 2`] = `
Object {
  "index": 0,
  "notation": "h",
  "type": "foundation",
  "value": Array [],
}
`;

exports[`notation-parser should parse moves to open cell 1`] = `
Object {
  "index": 0,
  "notation": "a",
  "type": "cell",
  "value": Array [],
}
`;

exports[`notation-parser should parse moves to open cell 2`] = `
Object {
  "index": 0,
  "notation": "1",
  "type": "tableau",
  "value": Array [
    Object {
      "color": "red",
      "id": "AD",
      "rank": "A",
      "suit": "D",
      "value": 1,
    },
    Object {
      "color": "red",
      "id": "QH",
      "rank": "Q",
      "suit": "H",
      "value": 12,
    },
    Object {
      "color": "black",
      "id": "QS",
      "rank": "Q",
      "suit": "S",
      "value": 12,
    },
    Object {
      "color": "black",
      "id": "6S",
      "rank": "6",
      "suit": "S",
      "value": 6,
    },
    Object {
      "color": "black",
      "id": "3C",
      "rank": "3",
      "suit": "C",
      "value": 3,
    },
    Object {
      "color": "black",
      "id": "2C",
      "rank": "2",
      "suit": "C",
      "value": 2,
    },
    Object {
      "color": "black",
      "id": "8S",
      "rank": "8",
      "suit": "S",
      "value": 8,
    },
  ],
}
`;
